scripted: akka-grpc server (scala) should pass the status_code_and_message integration test

Variation: akka-grpc server scala with grpc-java client tester

https://travis-ci.org/akka/akka-grpc/jobs/362603376

```
[info] Running test status_code_and_message
[info] [INFO] [04/05/2018 12:45:16.399] [pool-1-thread-1-ScalaTest-running-GrpcInteropSpec] [akka.actor.ActorSystemImpl(default)] Exception thrown, unbinding
[error] org.scalatest.exceptions.TestFailedException
[error] 	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:528)
[error] 	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:527)
[error] 	at org.scalatest.WordSpec.newAssertionFailedException(WordSpec.scala:1881)
[error] 	at org.scalatest.Assertions.fail(Assertions.scala:1121)
[error] 	at org.scalatest.Assertions.fail$(Assertions.scala:1117)
[error] 	at org.scalatest.WordSpec.fail(WordSpec.scala:1881)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.withGrpcServer(GrpcInteropTests.scala:74)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.$anonfun$grpcTests$4(GrpcInteropTests.scala:50)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.pendingTestCaseSupport(GrpcInteropTests.scala:86)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.$anonfun$grpcTests$3(GrpcInteropTests.scala:49)
[error] 	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[error] 	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[error] 0m	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[error] 	at org.scalatest.Transformer.apply(Transformer.scala:22)
[error] 	at org.scalatest.Transformer.apply(Transformer.scala:20)
[error] 	at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[error] 	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[error] 	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[error] 	at org.scalatest.WordSpec.withFixture(WordSpec.scala:1881)
[error] 	at org.scalatest.WordSpecLike.invokeWithFixture$1(WordSpecLike.scala:1076)
[error] 	at org.scalatest.WordSpecLike.$anonfun$runTest$1(WordSpecLike.scala:1088)
[error] 	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[error] 	at org.scalatest.WordSpecLike.runTest(WordSpecLike.scala:1088)
[error] 	at org.scalatest.WordSpecLike.runTest$(WordSpecLike.scala:1070)
[error] 	at org.scalatest.WordSpec.runTest(WordSpec.scala:1881)
[error] 	at org.scalatest.WordSpecLike.$anonfun$runTests$1(WordSpecLike.scala:1147)
[error] 	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
[error] 	at scala.collection.immutable.List.foreach(List.scala:389)
[error] 	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[error] 	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:373)
[error] 	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:410)
[error] 	at scala.collection.immutable.List.foreach(List.scala:389)
[error] 	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[error] 	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
[error] 	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[error] 	at org.scalatest.WordSpecLike.runTests(WordSpecLike.scala:1147)
[error] 	at org.scalatest.WordSpecLike.runTests$(WordSpecLike.scala:1146)
[error] 	at org.scalatest.WordSpec.runTests(WordSpec.scala:1881)
[error] 	at org.scalatest.Suite.run(Suite.scala:1147)
[error] 	at org.scalatest.Suite.run$(Suite.scala:1129)
[error] 	at org.scalatest.WordSpec.org$scalatest$WordSpecLike$$super$run(WordSpec.scala:1881)
[error] 	at org.scalatest.WordSpecLike.$anonfun$run$1(WordSpecLike.scala:1192)
[error] 	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[error] 	at org.scalatest.WordSpecLike.run(WordSpecLike.scala:1192)
[error] 	at org.scalatest.WordSpecLike.run$(WordSpecLike.scala:1190)
[error] 	at org.scalatest.WordSpec.run(WordSpec.scala:1881)
[error] 	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[error] 	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
[error] 	at sbt.ForkMain$Run$2.call(ForkMain.java:300)
[error] 	at sbt.ForkMain$Run$2.call(ForkMain.java:290)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] Caused by: java.lang.AssertionError: expected:<UNKNOWN> but was:<CANCELLED>
[error] 	at org.junit.Assert.fail(Assert.java:88)
[error] 	at org.junit.Assert.failNotEquals(Assert.java:834)
[error] 	at org.junit.Assert.assertEquals(Assert.java:118)
[error] 	at org.junit.Assert.assertEquals(Assert.java:144)
[error] 	at io.grpc.testing.integration.AbstractInteropTest.statusCodeAndMessage(AbstractInteropTest.java:1507)
[error] 	at io.grpc.testing.integration2.GrpcJavaClientTester.statusCodeAndMessage(GrpcJavaClientTester.java:126)
[error] 	at io.grpc.testing.integration2.TestServiceClient.runTest(TestServiceClient.java:186)
[error] 	at io.grpc.testing.integration2.TestServiceClient.run(TestServiceClient.java:93)
[error] 	at com.lightbend.grpc.interop.IoGrpcClient$.run(IoGrpcClient.scala:20)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.runGrpcClient(GrpcInteropTests.scala:80)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.$anonfun$grpcTests$5(GrpcInteropTests.scala:50)
[error] 	at com.lightbend.grpc.interop.GrpcInteropTests.withGrpcServer(GrpcInteropTests.scala:62)
[error] 	... 47 more
[info] [info] - should pass the status_code_and_message integration test *** FAILED ***
[info] [info]   org.scalatest.exceptions.TestFailedException was thrown. (GrpcInteropTests.scala:74)
```

--- a/interop-tests/src/test/scala/com/lightbend/grpc/interop/TestServiceImpl.scala
+++ b/interop-tests/src/test/scala/com/lightbend/grpc/interop/TestServiceImpl.scala
@@ -43,8 +43,8 @@ class TestServiceImpl(implicit ec: ExecutionContext, mat: Materializer) extends 
       case Some(requestStatus) =>
         val responseStatus = Status.fromCodeValue(requestStatus.code).withDescription(requestStatus.message)
         //  - Either one of the following works
-        // throw new GrpcServiceException(responseStatus)
-        Future.failed(throw new GrpcServiceException(responseStatus))
+        // Future.failed(new GrpcServiceException(responseStatus))
+        throw new GrpcServiceException(responseStatus)
     }
   }
 

--- a/runtime/src/main/scala/akka/grpc/internal/CancellationBarrierGraphStage.scala
+++ b/runtime/src/main/scala/akka/grpc/internal/CancellationBarrierGraphStage.scala
@@ -16,28 +16,26 @@ class CancellationBarrierGraphStage[T] extends GraphStage[FlowShape[T, T]] {
   override val shape: FlowShape[T, T] = FlowShape(in, out)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new GraphStageLogic(shape) with InHandler with OutHandler {
-      var finished = false
-      var pulled = false
+    new GraphStageLogic(shape) {
 
-      override def onPush(): Unit = {
-        pulled = false
-        val value = grab(in)
-        if (!finished)
-          emit(out, value)
-      }
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = emit(out, grab(in))
+      })
 
-      override def onPull(): Unit = {
-        pulled = true
-        pull(in)
-      }
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = pull(in)
 
-      override def onDownstreamFinish(): Unit = {
-        finished = true
-        if (!pulled)
-          pull(in)
-      }
+        override def onDownstreamFinish(): Unit = {
+          if (!hasBeenPulled(in))
+            pull(in)
 
-      setHandlers(in, out, this)
+          setHandler(in, new InHandler {
+            override def onPush(): Unit = {
+              grab(in)
+              pull(in)
+            }
+          })
+        }
+      })
     }
 }

--- a/runtime/src/main/scala/akka/grpc/internal/CancellationBarrierGraphStage.scala
+++ b/runtime/src/main/scala/akka/grpc/internal/CancellationBarrierGraphStage.scala
@@ -1,0 +1,43 @@
+package akka.grpc.internal
+
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+
+/**
+ * 'barrier' that makes sure that, even when downstream is cancelled,
+ * the complete upstream is consumed.
+ *
+ * @tparam T
+ */
+class CancellationBarrierGraphStage[T] extends GraphStage[FlowShape[T, T]] {
+  val in: Inlet[T] = Inlet("CancellationBarrier")
+  val out: Outlet[T] = Outlet("CancellationBarrier")
+
+  override val shape: FlowShape[T, T] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      var finished = false
+      var pulled = false
+
+      override def onPush(): Unit = {
+        pulled = false
+        val value = grab(in)
+        if (!finished)
+          emit(out, value)
+      }
+
+      override def onPull(): Unit = {
+        pulled = true
+        pull(in)
+      }
+
+      override def onDownstreamFinish(): Unit = {
+        finished = true
+        if (!pulled)
+          pull(in)
+      }
+
+      setHandlers(in, out, this)
+    }
+}

--- a/runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala
+++ b/runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala
@@ -8,15 +8,15 @@ import java.util.concurrent.{ CompletableFuture, CompletionStage }
 
 import io.grpc.Status
 import akka.NotUsed
-import akka.grpc.GrpcServiceException
-import akka.http.scaladsl.model.HttpEntity.{ ChunkStreamPart, LastChunk }
-import akka.http.scaladsl.model.{ HttpEntity ⇒ SHttpEntity, HttpResponse ⇒ SHttpResponse }
+import akka.http.scaladsl.model.HttpEntity.LastChunk
+import akka.http.scaladsl.model.{ HttpEntity => SHttpEntity, HttpResponse => SHttpResponse }
 import akka.http.scaladsl.model.headers.RawHeader
 import akka.http.javadsl.model.{ HttpRequest, HttpResponse }
 import akka.stream.Materializer
 import akka.stream.javadsl.{ Sink, Source }
-import akka.stream.scaladsl.{ Source ⇒ SSource }
+import akka.stream.scaladsl.{ Source => SSource }
 import akka.grpc._
+import akka.grpc.internal.CancellationBarrierGraphStage
 import akka.grpc.scaladsl.headers.`Message-Encoding`
 
 object GrpcMarshalling {
@@ -31,7 +31,10 @@ object GrpcMarshalling {
       req.entity.getDataBytes
         .mapMaterializedValue(_ ⇒ NotUsed)
         .via(Grpc.grpcFramingDecoder(messageEncoding))
-        .map(u.deserialize))
+        .map(u.deserialize)
+        // In gRPC we signal failure by returning an error code, so we
+        // don't want the cancellation bubbled out
+        .via(new CancellationBarrierGraphStage))
   }
 
   def marshal[T](e: T, m: ProtobufSerializer[T], mat: Materializer, codec: Codec): HttpResponse =

--- a/runtime/src/main/scala/akka/grpc/scaladsl/GrpcMarshalling.scala
+++ b/runtime/src/main/scala/akka/grpc/scaladsl/GrpcMarshalling.scala
@@ -6,11 +6,11 @@ package akka.grpc.scaladsl
 
 import akka.NotUsed
 import akka.grpc._
+import akka.grpc.internal.CancellationBarrierGraphStage
 import akka.grpc.scaladsl.headers.`Message-Encoding`
 import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
 import akka.stream.Materializer
 import akka.stream.scaladsl.{ Sink, Source }
-import akka.util.ByteString
 
 import scala.concurrent.Future
 
@@ -29,7 +29,10 @@ object GrpcMarshalling {
       req.entity.dataBytes
         .mapMaterializedValue(_ ⇒ NotUsed)
         .via(Grpc.grpcFramingDecoder(messageEncoding))
-        .map(u.deserialize))
+        .map(u.deserialize)
+        // In gRPC we signal failure by returning an error code, so we
+        // don't want the cancellation bubbled out
+        .via(new CancellationBarrierGraphStage))
   }
 
   def marshal[T](e: T = Identity)(implicit m: ProtobufSerializer[T], mat: Materializer, codec: Codec): HttpResponse =

